core/src/main/scala/org/http4s/headers/Forwarded.scala

```diff
@@ -0,0 +1,121 @@
+/*
+ * Copyright 2013-2020 http4s.org
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.http4s.headers
+
+import java.net.{Inet4Address, Inet6Address}
+
+import cats.data.NonEmptyList
+import org.http4s.parser.{ForwardedModelParsers, HttpHeaderParser}
+import org.http4s.util.Writer
+import org.http4s.{Header, HeaderKey, ParseResult, Uri}
+
+object Forwarded
+    extends HeaderKey.Internal[Forwarded]
+    with HeaderKey.Recurring
+    with ForwardedModelParsers {
+
+  final case class Node(nodeName: Node.Name, nodePort: Option[Node.Port] = None)
+
+  object Node {
+    def apply(nodeName: Name, nodePort: Port): Node = apply(nodeName, Some(nodePort))
+
+    sealed trait Name extends Any
+
+    object Name {
+      final case class Ipv4(address: Uri.Ipv4Address) extends AnyVal with Name
+      final case class Ipv6(address: Uri.Ipv6Address) extends AnyVal with Name
+      case object Unknown extends Name
+
+      def apply(address: Uri.Ipv4Address): Name = Name.Ipv4(address)
+      def apply(address: Inet4Address): Name = apply(Uri.Ipv4Address.fromInet4Address(address))
+      def apply(a: Byte, b: Byte, c: Byte, d: Byte): Name = apply(Uri.Ipv4Address(a, b, c, d))
+
+      def apply(address: Uri.Ipv6Address): Name = Name.Ipv6(address)
+      def apply(address: Inet6Address): Name = apply(Uri.Ipv6Address.fromInet6Address(address))
+      def apply(a: Short, b: Short, c: Short, d: Short, e: Short, f: Short, g: Short, h: Short)
+          : Name = apply(Uri.Ipv6Address(a, b, c, d, e, f, g, h))
+    }
+
+    sealed trait Port extends Any
+
+    object Port {
+      final case class Num(value: Int) extends AnyVal with Port
+
+      def apply(num: Int): Port = Port.Num(num)
+      def unapply(port: Port): Option[Int] =
+        PartialFunction.condOpt(port) {
+          case Num(num) => num
+        }
+    }
+
+    final case class Obfuscated(value: String) extends AnyVal with Name with Port
+
+    def fromString(s: String): ParseResult[Node] = new NodeParser(s).parse
+  }
+
+  final case class Host(host: Uri.Host, port: Option[Int] = None)
+
+  object Host {
+    def apply(host: Uri.Host, port: Int): Host = apply(host, Some(port))
+
+    def fromString(s: String): ParseResult[Host] = new HostParser(s).parse
+  }
+
+  type Proto = Uri.Scheme
+
+  sealed trait Element extends Any with Product {
+    def `by`: Option[Node]
+    def `for`: Option[Node]
+    def `host`: Option[Host]
+    def `proto`: Option[Proto]
+
+    def `by=`(value: Node): Element
+    def `for=`(value: Node): Element
+    def `host=`(value: Host): Element
+    def `proto=`(value: Proto): Element
```

Comment:
We use `withFoo` instead of `foo=` elsewhere.  Maybe that's better here?

Reply:
Yeah, I anticipated you won't like it:)
Actually, I was just playing a bit with different approaches.
I agree – it's better to keep this DSL consistent with others parts.

Reply:
It reminded me of Scala's `_=` operator syntax, but those are for side effects.  If your convention didn't require the backticks, I'd probably like it a lot! :smile: